lib/ddtrace/contrib/excon/patcher.rb

```diff
@@ -0,0 +1,58 @@
+module Datadog
+  module Contrib
+    module Excon
+      COMPATIBLE_UNTIL = Gem::Version.new('1.0.0')
+      SERVICE = 'excon-request'.freeze
+
+      # Responsible for hooking the instrumentation into excon
+      module Patcher
+        include Base
+        register_as :excon, auto_patch: true
+        option :distributed_tracing_enabled, default: false
+        option :split_by_domain, default: false
+        option :error_handler, default: nil
+
+        @patched = false
+
+        class << self
+          def patch
+            return @patched if patched? || !compatible?
+
+            require 'ddtrace/ext/app_types'
+            require 'ddtrace/contrib/excon/middleware'
+
+            add_pin
+            add_middleware
+
+            @patched = true
+          rescue => e
+            Tracer.log.error("Unable to apply Excon integration: #{e}")
+            @patched
+          end
+
+          def patched?
+            @patched
+          end
+
+          private
+
+          def compatible?
+            return unless defined?(::Excon::VERSION)
+
+            Gem::Version.new(::Excon::VERSION) < COMPATIBLE_UNTIL
+          end
+
+          def add_pin
+            Pin.new(SERVICE, app_type: Ext::AppTypes::WEB).tap do |pin|
+              pin.onto(::Excon)
+            end
+          end
+
+          def add_middleware
+            ::Excon.defaults[:middlewares].append(Middleware)
```

Comment:
so, I tracked the weirdness i was seeing in our excon instrumentation to this middleware being appended after another middleware that made some redis calls.  Hence the order was messed up so it looked async.  I think maybe this should be prepend, but not sure, or what other ordering issues may come up from this

Reply:
I think the best place to do this is by injecting the trace middleware as soon as possible in the stack, but behind the Response middleware.

With some code changes, you can do either of these:

```ruby
# Default middleware stack
# This is what the patcher now does
Excon.new(
  'http://example.com',
  middlewares: Datadog::Contrib::Excon::Middleware.with(middleware_options).around_default_stack
)

# Custom middleware stack
Excon.new(
  'http://example.com',
  middlewares: [
    Excon::Middleware::ResponseParser,
    Datadog::Contrib::Excon::Middleware.with(middleware_options),
    Excon::Middleware::Mock
  ]
)
```